lib/backends/cpu/ops/matmul.ts

```diff
@@ -16,71 +14,153 @@ export class CpuMatMul extends MatMul {
+  const shape = BroadcastUtil.calcShape(dimsA, dimsB, true);
   if (!shape) {
     // the inputs cannot broadcast or cannot multiply
     throw new Error(`input dimensions do not match the requirement`);
   }
-  // make a copy and re-assign because this can be modified later
-  shape = shape.slice(0);
   const size = ShapeUtil.size(shape);
   const num2dMatrices = size / (mat2dShape[0] * mat2dShape[1]);
 
-  let ndA: ndarray;
-  let ndB: ndarray;
-  let ndY: ndarray;
-  let isFloat64 = false;
-  ndA = ndarray(a.floatData, dimsA);
-  ndB = ndarray(b.floatData, dimsB);
-  if (a.type === 'float64' || b.type === 'float64') {
-    ndY = ndarray(new Float64Array(size));
-    isFloat64 = true;
-  } else {
-    ndY = ndarray(new Float32Array(size));
-  }
-
-  let curPos = 0;
+  const y = new Tensor(shape, a.type === 'float64' || b.type === 'float64' ? 'float64' : 'float32');
+  let offsetY = 0;
   const indices = new Array<number>(shape.length);
-  const indicesA = new Array(ndA.shape.length);
-  const indicesB = new Array(ndB.shape.length);
+  const indicesA = new Array<number>(a.dims.length);
+  const indicesB = new Array<number>(b.dims.length);
   for (let i = 0; i < num2dMatrices; i++) {
     // traverse nd array at 2d level
+    indices[shape.length - 2] = 0;
+    indices[shape.length - 1] = 0;
     let rest = i;
     for (let j = shape.length - 3; j >= 0; j--) {
       indices[j] = rest % shape[j];
       rest = Math.floor(rest / shape[j]);
     }
-    // map the "broadcasted" index to original ndarray index
-    BroadcastUtil.fillIndex(indices, ndA.shape, indicesA);
-    BroadcastUtil.fillIndex(indices, ndB.shape, indicesB);
-    // slice and get 2d subarrays
-    const subarrayA = shape.length === 2 ? ndA : ndA.pick(...indicesA);
-    const subarrayB = shape.length === 2 ? ndB : ndB.pick(...indicesB);
+    // map the "broadcasted" index to original index
+    BroadcastUtil.fillIndex(indices, a.dims, indicesA);
+    BroadcastUtil.fillIndex(indices, b.dims, indicesB);
+    // calculate subarrays offset for A and B
+    const offsetA = indicesA.length <= 2 ? 0 : ShapeUtil.indicesToOffset(indicesA, a.strides, shape.length - 2);
+    const offsetB = indicesB.length <= 2 ? 0 : ShapeUtil.indicesToOffset(indicesB, b.strides, shape.length - 2);
     // multiply like conventional matrices
-    MatMul2d(subarrayA, subarrayB, ndY, curPos);
-    curPos += mat2dShape[0] * mat2dShape[1];
+    matMul2d(
+        a.floatData.subarray(offsetA), b.floatData.subarray(offsetB), y.floatData.subarray(offsetY), false, false, 1, 0,
+        mat2dShape[0], mat2dShape[1], dimsA[dimsA.length - 1]);
+    offsetY += mat2dShape[0] * mat2dShape[1];
+  }
+  return y;
+}
+
+/**
+ * perform matrix multiply on C = alpha * A * B + beta * C
+ * @param A data of tensor A, whose shape is [M,K] or [K,M] (if transA)
+ * @param B data of tensor B, whose shape is [K,N] or [N,K] (if transB)
+ * @param C data of tensor C, whose shape is [M,N]
+ */
+export function matMul2d(
```

Comment:
Are these helpers coming from Lotus ? Should we check-in some explicit unit tests for these (can come in a separate PR). Currently, I guess they are being implicitly tested through model usage.